### PR TITLE
APEXCORE-495 supporting apps in config package.

### PR DESCRIPTION
**Pull request by sandeshh**
1. Added support to store apps in the config file
2. By default apps from the config and app package gets merged (During conflict Config apps overwrite the AppPackage apps)
3. New option during launch, "useConfigApps" - will take either of "inclusive"/"exclusive"
   "inclusive" - merge the apps present in app & config package and the config package apps overwrite the conflicting the apps
   "exclusive" - show only config apps
   If the user do not specify "useConfigApps" only the apps present in the app package is shown.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 1]
```diff
@@ -3715,6 +3718,50 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+
+    if (!src.exists()) {
+      return ap.getApplications();
+    }
+
+    if (onlyConfigApps) {
+      try {
+        FileUtils.cleanDirectory(dest);
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Need appropriate exception handling

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 2]
```diff
@@ -3715,6 +3718,50 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+      for (File app: src.listFiles()) {
+        FileUtils.deleteQuietly(new File(dest, app.getName()));
+      }
+    }
+
+    for (File app: src.listFiles()) {
+      try {
+        FileUtils.moveFileToDirectory(app, dest, true);
+      } catch (IOException e) {
+        LOG.error("Application from the config file {} failed while processing", app.getName());
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Should it be a warn since it's swallowing the exception and moving on?

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 3]
```diff
@@ -3715,6 +3718,50 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+      }
+    }
+
+    try {
+      FileUtils.deleteDirectory(src);
+    } catch (IOException e) {
+      LOG.error("Failed to delete the Config Apps folder");
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Ditto above

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 4]
```diff
@@ -3850,6 +3897,7 @@ private Option add(Option opt)
     final Option exactMatch = add(new Option("exactMatch", "Only consider applications with exact app name"));
     final Option queue = add(OptionBuilder.withArgName("queue name").hasArg().withDescription("Specify the queue to launch the application").create("queue"));
     final Option force = add(new Option("force", "Force launch the application. Do not check for compatibility"));
+    final Option onlyConfigApps = add(new Option("onlyConfigApps","Only use the apps present in the config file."));
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: "config package file". We should make it very clear that it's different from the -conf parameter, which takes a config xml file.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 5]
```diff
@@ -371,23 +376,12 @@ private void processAppDirectory(File dir)
...
-          } catch (Exception ex) {
-            appInfo.error = ex.getMessage();
-            appInfo.errorStackTrace = ExceptionUtils.getStackTrace(ex);
-          }
+        AppInfo appInfo = jsonFileToAppInfo(entry,config);
+
+        if (appInfo != null) {
           applications.add(appInfo);
-        } catch (Exception ex) {
-          LOG.error("Caught exceptions trying to process {}", entry.getName(), ex);
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: Ditto above

File: engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java [hunk 6]
```diff
@@ -37,10 +37,13 @@
 
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
 
 import net.lingala.zip4j.core.ZipFile;
 import net.lingala.zip4j.exception.ZipException;
 
+import static com.datatorrent.stram.client.StramClientUtils.jsonFileToAppInfo;
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java]: jsonFileToAppInfo is only used in one place. We should probably just use StramClientUtils.jsonFileToAppInfo, instead of import static here, to be consistent with the rest of StramClientUtils calls.

File: engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java [hunk 7]
```diff
@@ -49,6 +49,7 @@
 import org.slf4j.LoggerFactory;
 
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.exception.ExceptionUtils;
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java]: We should use lang3 whenever possible.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 8]
```diff
@@ -3715,6 +3718,50 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+    }
+
+    if (onlyConfigApps) {
+      try {
+        FileUtils.cleanDirectory(dest);
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    } else {
+      for (File app: src.listFiles()) {
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Please check the style of your IDE. There should be a space before this colon. Also check all other occurrences.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 9]
```diff
@@ -75,6 +75,7 @@
 import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.shell.Command;
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Why this import?

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java (hunk 4 above)
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Also, check the style rule. There should be a space after the comma.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 10]
```diff
@@ -3850,6 +3908,9 @@ private Option add(Option opt)
     final Option exactMatch = add(new Option("exactMatch", "Only consider applications with exact app name"));
     final Option queue = add(OptionBuilder.withArgName("queue name").hasArg().withDescription("Specify the queue to launch the application").create("queue"));
     final Option force = add(new Option("force", "Force launch the application. Do not check for compatibility"));
+    final Option onlyConfigApps = add(new Option("onlyConfigApps", "Only use the apps present in the config package."));
+    final Option skipConfigApps = add(new Option("skipConfigApps", "Skip the apps present in the config package."));
+
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Instead of having two options that conflict each other, why don't we have one option that takes a value? for example --useConfigPackageApps={exclusive|inclusive|none}

File: engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java [hunk 11]
```diff
@@ -71,6 +72,7 @@
 
   private final Map<String, String> properties = new TreeMap<>();
   private final Map<String, Map<String, String>> appProperties = new TreeMap<>();
+  private List<AppPackage.AppInfo> applications = new ArrayList<>();
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java]: final?

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java (hunk 10 above)
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Yes, that is better. I will change it.

File: engine/src/test/java/com/datatorrent/stram/support/StramTestSupport.java [hunk 12]
```diff
@@ -157,21 +160,42 @@ public static void waitForWindowComplete(OperatorContext nodeCtx, long windowId)
    */
   public static File createAppPackageFile()
   {
+    buildOperators();
+
     final String version = System.getProperty("apex.version");
     final List<String> params = new LinkedList<>();
     params.add("clean");
-    params.add("package");
+    params.add("install");
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/support/StramTestSupport.java]: why this change?

File: engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/RandomGen.java [hunk 13]
```diff
@@ -16,24 +16,44 @@
...
 import com.datatorrent.api.InputOperator;
 import com.datatorrent.common.util.BaseOperator;
 
-/**
- * This is a simple operator that emits random number
- */
-
-public class RandomNumberGenerator extends BaseOperator implements InputOperator
+public class RandomGen extends BaseOperator implements InputOperator
 {
```
Comment (davidyan74) [inline engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/RandomGen.java]: why this change?

File: engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/DevNull.java [hunk 14]
```diff
@@ -16,25 +16,19 @@
...
 
-/**
- * This is a simple output operator that prints everything to stdout
- */
-
-public class StdoutOperator extends BaseOperator
+public class DevNull extends BaseOperator
 {
-  public final transient DefaultInputPort<Object> in = new DefaultInputPort<Object>()
+  public transient DefaultInputPort<Object> inputPort = new DefaultInputPort<Object>()
```
Comment (davidyan74) [inline engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/DevNull.java]: why this change?

File: engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java (hunk 11 above)
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/client/ConfigPackage.java]: Yes, it can be final

File: engine/src/test/java/com/datatorrent/stram/support/StramTestSupport.java (hunk 12 above)
Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/support/StramTestSupport.java]: will make it package

File: engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/RandomGen.java (hunk 13 above)
Comment (sandeshh) [inline engine/src/test/resources/testOperators/src/main/java/org/apache/apex/test/RandomGen.java]: I had these operators created and tested separately, then deleted the old ones.

**Comment by sandeshh** [comment]
@davidyan74 Please review.

File: apex-conf-archetype/src/main/resources/archetype-resources/pom.xml [hunk 15]
```diff
@@ -25,6 +25,7 @@
     <apex.apppackage.maxversion>1.9999.9999</apex.apppackage.maxversion>
     <apex.appconf.classpath>classpath/*</apex.appconf.classpath>
     <apex.appconf.files>files/*</apex.appconf.files>
+    <apex.appconf.files>app/*</apex.appconf.files>
```
Comment (davidyan74) [inline apex-conf-archetype/src/main/resources/archetype-resources/pom.xml]: duplicate xml tag apex.appconf.files.

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 16]
```diff
@@ -49,6 +49,8 @@
 import net.lingala.zip4j.exception.ZipException;
 import net.lingala.zip4j.model.ZipParameters;
 
+import static com.datatorrent.stram.client.StramClientUtils.jsonFileToAppInfo;
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: Please just use StramClientUtils.jsonFileToAppInfo instead of having import static.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java [hunk 17]
```diff
@@ -180,4 +188,40 @@ public void testLaunchAppPackagePrecedenceWithConfigPackage() throws Exception
...
+        ApexCli.getLaunchCommandLineInfo(new String[]{"-conf", configFile.getAbsolutePath(), appFile.getAbsolutePath(), "-useConfigApps", "inclusive"});
+
+    Assert.assertEquals("configApps", "inclusive", commandLineInfo.useConfigApps);
+
+    ApexCli apexCli = new ApexCli();
+    apexCli.init();
+
+    try {
+      apexCli.getLaunchAppPackageArgs(ap, cp, commandLineInfo, new ConsoleReader());
+    } catch (Throwable throwable ) {
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: why are we catching throwable instead of the exception you expect it to throw? Also, there is an extra space before the closing paren.

Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: Also, why should this throw an exception? It should be okay to have the conf package to include multiple apps and in which case, let the user choose by showing a menu, just like app package can contain multiple apps.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java [hunk 18]
```diff
@@ -103,8 +107,8 @@ public static void finished()
   public void testLaunchAppPackagePropertyPrecedence() throws Exception
   {
     // set launch command options
-    ApexCli.LaunchCommandLineInfo commandLineInfo = ApexCli
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java:106]: Why are we making such changes and have an extra import static? Please restrict your changes to **only** those that are relevant to this feature.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java [hunk 19]
```diff
@@ -126,10 +130,14 @@ public void testLaunchAppPackagePropertyPrecedence() throws Exception
   @Test
   public void testLaunchAppPackageParametersWithConfigPackage() throws Exception
   {
-    ApexCli.LaunchCommandLineInfo commandLineInfo = ApexCli
-        .getLaunchCommandLineInfo(new String[]{"-exactMatch", "-conf", configFile.getAbsolutePath(), appFile.getAbsolutePath(), "MyFirstApplication"});
+    ApexCli.LaunchCommandLineInfo commandLineInfo =
+        getLaunchCommandLineInfo(new String[]{"-exactMatch", "-conf", configFile.getAbsolutePath(), appFile.getAbsolutePath(), "MyFirstApplication"});
+
+    commandLineInfo.args = new String[1];
+    commandLineInfo.args[0] = "MyFirstApplication";
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: Please use the notation `new String[]{ "item1", "item2", ... }` instead of doing element assignment like this.

File: engine/src/test/resources/testAppPackage/mydtapp/src/main/resources/com/example/mydtapp/Application/ExtRes.html [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<html>example</html>
```
Comment (davidyan74) [inline engine/src/test/resources/testAppPackage/mydtapp/src/main/resources/com/example/mydtapp/Application/ExtRes.html]: What is this file for?

File: engine/src/test/resources/testConfigPackage/testConfigPackageSrc/app/testApp.json [hunk 21]
```diff
@@ -0,0 +1 @@
+{"displayName":"testApp","attributes":{},"operators":[{"name":"Operator 1","attributes":{},"class":"org.apache.apex.test.DevNull","ports":[{"name":"inputPort","attributes":{}}]},{"name":"Operator 2","attributes":{},"class":"org.apache.apex.test.RandomGen","ports":[{"name":"out","attributes":{}}]}],"streams":[{"name":"Stream 1","sinks":[{"operatorName":"Operator 1","portName":"inputPort"}],"source":{"operatorName":"Operator 2","portName":"out"}}]}
```
Comment (davidyan74) [inline engine/src/test/resources/testConfigPackage/testConfigPackageSrc/app/testApp.json]: Please format the json file appropriately.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java (hunk 17 above)
Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: It throws, CLIException, but it is a private class. So I will convert it to catch RuntimeException.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java [hunk 22]
```diff
@@ -53,16 +56,16 @@
...
 
-  static Map<String, String> env = new HashMap<String, String>();
-  static String userHome;
+  Map<String, String> env = new HashMap<String, String>();
+  String userHome;
 
-  @BeforeClass
-  public static void starting()
+  @Before
+  public void starting()
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: Can you explain why this change is needed?

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java (hunk 17 above)
Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: Feature, you mentioned is already present. Exception is also an existing behaviour, if the apex cli is not connected to console it throws an exception. In the unit test there is no console.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java (hunk 22 above)
Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: The merging of the apps deletes the apps from the app-packages. It means that the extracted packages cannot be reused. So setup needs to be done before every test. I do see some operation that can be done at class level.

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java (hunk 17 above)
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: OK. Please just change the catch throwable then

Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: BTW, you can make the CLIException package scope so that you can use it in unit tests.

Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java]: will do.

**Comment by sandeshh** [comment]
@davidyan74 All the review comments have been addressed.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 23]
```diff
@@ -3715,6 +3720,52 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+    if (!src.exists()) {
+      return ap.getApplications();
+    }
+
+    if (configApps.equals(CONFIG_EXCLUSIVE)) {
+
+      for (File app : dest.listFiles()) {
+
+        if (app.getName().endsWith("json")) {
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: endsWith(".json")

Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: also, change the variable name to "file" to avoid confusion.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 24]
```diff
@@ -3715,6 +3720,52 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+          FileUtils.deleteQuietly(new File(dest, app.getName()));
+        }
+      }
+    } else {
+      for (File app : src.listFiles()) {
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: ditto above

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 25]
```diff
@@ -3715,6 +3720,52 @@ DTConfiguration getLaunchAppPackageProperties(AppPackage ap, ConfigPackage cp, L
...
+        FileUtils.deleteQuietly(new File(dest, app.getName()));
+      }
+    }
+
+    for (File app : src.listFiles()) {
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: ditto above

**Comment by davidyan74** [comment]
@sandeshh looks pretty good now. please address the last nitpick. I'll merge it when that's done.

**Comment by sandeshh** [comment]
@davidyan74 Done!!!

File: engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java [hunk 26]
```diff
@@ -180,4 +204,40 @@ public void testLaunchAppPackagePrecedenceWithConfigPackage() throws Exception
...
+        .getLaunchCommandLineInfo(new String[]{"-conf", configFile.getAbsolutePath(), appFile.getAbsolutePath(), "-useConfigApps", "exclusive"});
+
+    ApexCli apexCli = new ApexCli();
+    apexCli.init();
+
+    Assert.assertEquals("configApps", "exclusive", commandLineInfo.useConfigApps);
+
+    apexCli.getLaunchAppPackageArgs(ap, cp, commandLineInfo, new ConsoleReader());
+
+    Assert.assertEquals(ap.getApplications().size(), 1);
```
Comment (davidyan74) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java:221]: can you add verification of the only element of the application list? should check that the name matches and that it's a json app.

Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/cli/ApexCliTest.java:221]: Done
